### PR TITLE
Modularize wails

### PR DESCRIPTION
**Pull request by leaanthony**
This PR aims to make a Wails project more configurable as requested in #1877.
Changes:
  - A new `frontend:dir` entry in `wails.json`. This can be an absolute path or a relative path from the `wails.json` file. The last segment of this path is considered the name of the frontend directory. Not specifying this will result in the surrent default behaviour.
  - If `go.mod` is not found in the same directory as `wails.json`, the cli will look up the parent directories to find it.

Note: The CLI commands `build` and `dev` must still be run in the same directory as `wails.json`.

Example project in `v2/examples/customlayout`

### CODE REVIEW

File: v2/internal/project/project.go [hunk 1]
```diff
@@ -36,7 +37,7 @@ type Project struct {
 	/*** Internal Data ***/
 
 	// The path to the project directory
-	Path string
+	Path string `json:"projectdir"`
```
Comment (stffabi) [inline v2/internal/project/project.go:40]: We are now exposing this to the user, what's the meaning of this folder? AFAIK it's the folder where the `main.go` could be found, should we use a better naming for this?
Should this also be relative to the location of `wails.json` like frontendir and wailsjsdir? This would make it possible to put wails.json in the root of a project and put main.go of wails in e.g. `cmd/desktop`.

File: v2/internal/project/project.go [hunk 2]
```diff
@@ -88,6 +89,23 @@ type Project struct {
...
+
+	// Frontend directory
+	FrontendDir string `json:"frontend:dir"`
+}
+
+func (p *Project) GetFrontendDir() string {
+	if filepath.IsAbs(p.FrontendDir) {
+		return p.FrontendDir
+	}
+	return filepath.Join(p.Path, p.FrontendDir)
```
Comment (stffabi) [inline v2/internal/project/project.go:101]: Since we are exposing `p.Path` to the user `p.FrontendDir` is not always relative to where `wails.json` is found. It might be relative to that or relative to `projectdir`.
Would it be more consistent and easier to reason about this folder if it is always relative to `wails.json` and `projectdir` is also relative to `wails.json`?
So every folder defined in `wails.json` would be relative to the location if the file itself.

File: v2/internal/project/project.go [hunk 3]
```diff
@@ -88,6 +89,23 @@ type Project struct {
...
+}
+
+func (p *Project) GetWailsJSDir() string {
+	if filepath.IsAbs(p.WailsJSDir) {
+		return p.WailsJSDir
+	}
+	return filepath.Join(p.Path, p.WailsJSDir)
```
Comment (stffabi) [inline v2/internal/project/project.go:108]: See comment on `GetFrontendDir`

File: v2/internal/project/project.go [hunk 4]
```diff
@@ -119,6 +137,73 @@ func (p *Project) Save() error {
...
+	}
+	if p.Info.Copyright == nil {
+		v := "Copyright........."
+		p.Info.Copyright = &v
+	}
+	if p.Info.Comments == nil {
+		v := "Built using Wails (https://wails.io)"
+		p.Info.Comments = &v
+	}
+	if p.DevServer == "" {
```
Comment (stffabi) [inline v2/internal/project/project.go:200]: Seems like a duplicate of line 170.

File: v2/internal/project/project.go [hunk 5]
```diff
@@ -119,6 +137,73 @@ func (p *Project) Save() error {
...
+		p.WailsJSDir = p.FrontendDir
+	}
+	if p.DebounceMS == 0 {
+		p.DebounceMS = 100
+	}
+	if p.DevServer == "" {
+		p.DevServer = "localhost:34115"
+	}
+	if p.NSISType == "" {
+		p.NSISType = "portable"
```
Comment (stffabi) [inline v2/internal/project/project.go:181]: IIRC portable is not a valid value for NSISType, it should be one of "single"/"multiple" and the default was empty which is mapped to "multiple".

```suggestion
		p.NSISType = "multiple"
```

File: v2/internal/project/project.go [hunk 6]
```diff
@@ -119,6 +137,73 @@ func (p *Project) Save() error {
...
+	}
+	if p.Version == "" {
+		p.Version = "2"
+	}
+	// Create default name if not given
+	if p.Name == "" {
+		p.Name = "wailsapp"
+	}
+	if p.AssetDirectory == "" {
+		p.AssetDirectory = "assets"
```
Comment (stffabi) [inline v2/internal/project/project.go:160]: Won't this break all projects that don't use a frontend devserver and rely on inferring of the AssetDirectory? The default until now was empty, which would be to infer it from the assets embed.FS.

**Review by stffabi — COMMENTED** [review]
Great enhancement 🚀 , left some minor comments.

File: v2/pkg/buildassets/buildassets.go [hunk 7]
```diff
@@ -83,6 +80,7 @@ func ReadFileWithProjectData(projectData *project.Project, file string) ([]byte,
 // ProjectInfo if necessary.
 // It will also write the resolved final file back to the project build folder.
 func ReadOriginalFileWithProjectDataAndSave(projectData *project.Project, file string) ([]byte, error) {
+	rootFolder := filepath.Base(projectData.GetBuildDir())
 	file = path.Join(rootFolder, file)
 	content, err := iofs.ReadFile(assets, file)
```
Comment (stffabi) [inline v2/pkg/buildassets/buildassets.go:85]: Same comment as above, I would say this should be: `content, err := iofs.ReadFile(assets, path.Join("build", file))` where file is the unchanged file argument.

File: v2/pkg/buildassets/buildassets.go [hunk 8]
```diff
@@ -124,15 +122,16 @@ func resolveProjectData(content []byte, projectData *project.Project) ([]byte, e
 }
 
 func writeFileSystemFile(projectData *project.Project, file string, content []byte) error {
-	path := filepath.Clean(filepath.Join(projectData.Path, filepath.FromSlash(file)))
+	baseDir := filepath.Dir(projectData.GetBuildDir())
+	targetPath := filepath.Clean(filepath.Join(baseDir, filepath.FromSlash(file)))
```
Comment (stffabi) [inline v2/pkg/buildassets/buildassets.go:126]: With the above changed code, file should always be a slash path relative to BuildDir, in which case the targetPath could just be: `targetPath := filepath.Join(projectData.GetBuildDir(), filepath.FromSlash(file))`

File: v2/pkg/commands/build/desktop.go [hunk 9]
```diff
@@ -20,9 +20,10 @@ func newDesktopBuilder(options *Options) *DesktopBuilder {
 func (d *DesktopBuilder) BuildAssets(options *Options) error {
 
 	// Check assets directory exists
-	if !fs.DirExists(options.ProjectData.BuildDir) {
+	buildDir := options.ProjectData.GetBuildDir()
+	if !fs.DirExists(buildDir) {
 		// Path to default assets
-		err := buildassets.Install(options.ProjectData.Path)
+		err := buildassets.Install(buildDir)
```
Comment (stffabi) [inline v2/pkg/commands/build/desktop.go:26]: Doesn't `buildassets.Install` already add a `build` dir to the path since the `assets embed.FS` from build dir contains that root folder?

File: v2/internal/project/project.go (hunk 1 above)
Comment (leaanthony) [reply v2/internal/project/project.go:40]: Yeah, this looks like it was left over from a previous branch I cherry picked :/ I'll take this out.

Comment (stffabi) [reply v2/internal/project/project.go:40]: Great, thanks for clarifying. So then some of my comments will become obsolete 😀

File: v2/internal/project/project.go (hunk 2 above)
Comment (leaanthony) [reply v2/internal/project/project.go:101]: Yes, I agree - every relative path in `wails.json`. Right now `p.Path` *is* where `wails.json` is (after I remove "projectdir") so I think it's good for now. I think eventually, `project.Load()` will be path-less and use cwd by default. It will traverse the filesystem to find it. Once found, it will load and write this location to the project data internally. This method will calculate based off that.

File: v2/internal/project/project.go (hunk 5 above)
Comment (leaanthony) [reply v2/internal/project/project.go:181]: WTF copilot! Oooph. Was this even set before?

File: v2/pkg/commands/build/desktop.go (hunk 9 above)
Comment (leaanthony) [reply v2/pkg/commands/build/desktop.go:26]: It does not, though I can see why it looks like it would. Running `wails build` in `examples/customlayout/cmd/customlayout` works as expected.

**Comment by leaanthony** [comment]
I feel a CLI rewrite coming on 😜

File: v2/internal/project/project.go (hunk 2 above)
Comment (stffabi) [reply v2/internal/project/project.go:101]: Yeah, with removing projectdir that's fine.

File: v2/internal/project/project.go (hunk 5 above)
Comment (stffabi) [reply v2/internal/project/project.go]: It was empty, which mapped to multiple.

File: v2/pkg/commands/build/desktop.go (hunk 9 above)
Comment (stffabi) [reply v2/pkg/commands/build/desktop.go:26]: That's interesting, I'm confused on that with diff of the PR. Before we passed e.g. `options.ProjectData.Path = ./` and it wrote it into `./build/`, now we are passing `./build/` and it also writes it into `./build/`.

Does that method `desktopBuilder.BuildAssets` even get called?

**Comment by stffabi** [comment]
Just added a commit which should fix the buildassets comments.

**Comment by mholt** [comment]
Hey, this works for what I need! :partying_face: 

I only tested it briefly, thanks for the great instructions. But so far it's quite nice.

**Comment by leaanthony** [comment]
Thanks for pushing for this. It really has been on the backburner for quite some time so good to get it done finally 🙏

**Comment by mholt** [comment]
Thank you for working on it. Wow, this was not a small change!

**Comment by leaanthony** [comment]
@mholt @stffabi - you think it's worth finishing this off with a `bin:dir` too or are we good with `${build:dir}/bin`?

File: v2/internal/project/project.go (hunk 6 above)
Comment (stffabi) [reply v2/internal/project/project.go:160]: Just saw this conversation was marked as resolved. I think we should still remove this default value to prevent problems for those users.

**Comment by stffabi** [comment]
I think we are good with `${build:dir}/bin` for the time being.

File: v2/internal/project/project.go (hunk 6 above)
Comment (leaanthony) [reply v2/internal/project/project.go:160]: Not sure if it's just threading issues on GitHub mobile but your reply is to a thread that's unresolved.

**Comment by leaanthony** [comment]
Thanks for explaining. Yeah, should use same build dir 👍

**Comment by mholt** [comment]
One thing I just thought of that I am curious your thoughts on, is that if I do change my `frontend:dir` (let's say, I'm testing something real quick) I also need to update my `go:embed` tag on my `embed.FS` value. If these are out of sync things get weird real quick.

It could be nice if Wails could infer the frontend dir from the `go:embed` on the fs.FS set on `options.App.Assets` (soon to be `options.App.AssetServer.Assets`), but since `Assets` takes an `fs.FS` and not an `embed.FS`, I guess there's no guarantee it's a real folder in the first place.

So in summary, I don't know an easy way to infer the frontend dir, and this is not a big deal to me, and definitely not a blocker for this PR or the next release. But I wanted to document this nuance here for the record.

Thanks again for this super handy improvement!

**Comment by stffabi** [comment]
It's pretty hard to infer that, since it is currently needed pretty early in the build process. So one would either need to do static analysis or build the binary in some special mode and execute it to get the inferred directory if the assets are an embed.FS. The problem is also how to find the root directory when inferring.

IIRC the frontend dir is also not exactly the same as the assets dir since the assets dir would point to the output dir of the frontend, e.g. frontend/dist.

I wonder if we could change that to some code generation method. We could add the assetdir again to wails.json which is the only source of truth for it. The build command could then create a 'assets.go' where 'main.go' is and with the correct go embed content.
'
//go:embed all:frontend/dist
var assets embed.FS
'

Would only work if the frontend dir is a child of the project dir. If it's in another dir one could add the embed there and add an import in the assts.go in the main package and referring to the assets in the other package.

Just a quick thought without thinking too much about the details and consequences...

**Comment by mholt** [comment]
@leaanthony Yep, totally understand. No problem.

@stffabi Ah yeah, you're right... I forget that `frontend` and `frontend/dist` ("Assets") are different for most people, as I have a vanilla JS frontend that doesn't use build steps. So to me they are the same.

Anyway, carry on. :smile:

**Comment by leaanthony** [comment]
> @leaanthony The dev build from `wails dev` is appearing in the folder from which I run `wails dev` (in my case, `cmd/timelinize`). It has the binary in it and then after I exit the program it goes away but the `build/bin` folders remain, empty.
> 
> I think they're supposed to be in the root of the project, right? (the build dir is at the root in the template project, at least)

@mholt - I am unable to reproduce this! The `dev` exe is appearing in the same place as the production binary for me with the `-dev` suffix 😖

**Comment by mholt** [comment]
@leaanthony Ok, no worries then! I probably misconfigured something in my setup. I wouldn't worry about it right now or let it be a blocker.

**Comment by leaanthony** [comment]
No worries @mholt. Just let me know if you can reproduce with a clean project and we'll fix it!

**Comment by olup** [comment]
Just wanted to add a useless comment sharing how exciting this is, I can't wait to finally refactor my app to include all sub modules alongside the wails code !

**Review by stffabi — APPROVED** [review]
LGTM 🚀

**Comment by W1M0R** [comment]
Just leaving this note here for others that might have the same issue.

The `../` path does not seem be supported by `go:embed`. So your frontend will still have to live next to or below the `wails.json` file:

```go
// THE FOLLOWING WON'T WORK

//go:embed all:../frontend/dist
var assets embed.FS
```
